### ISSUE TITLE
Auto-set numeric datatypes

### ISSUE THREAD

**Issue by JoschD** [pylhc/tfs#24]
Problem: 
When writing a dataframe without specifying the dtype beforehand one often runs into the error 
`unkown format '%s' for object of type 'int'` or similar, because `dtype=object`, the default if pandas does not know the datatype, maps to '%s' in our writer.

I would not automatically try to solve it, but I would add an easy solution. Talked to Lukas: Have it as a function in tools, then you can also use it on pandas DataFrames.  (please share opinion @Mael-Le-Garrec @lmalina @mihofer @awegsche 
)

```python
from functools import partial
import pandas as pd
(...)
def auto_dtype(df):
    return df.apply(partial(pd.to_numeric, errors='ignore')) 
```

**Comment by JoschD** [comment]
Example how this happens:
```python
import tfs
import numpy as np

df = tfs.TfsDataFrame(data=np.array([["A",1 ,2], ["B", 3, 4]], dtype=object), columns=["NAME", "BETX", "ALFX"])
tfs.write("test.tfs", df)
```

**Comment by fsoubelet** [comment]
@JoschD 

This would only be for `pandas >= 1.0`, but it seems the newly added [`convert_dtypes`](https://pandas.pydata.org/pandas-docs/version/1.0.0/reference/api/pandas.DataFrame.convert_dtypes.html#pandas.DataFrame.convert_dtypes) could make the problem disappear in most cases? We probably don't deal with types that would confuse it, and it rids us of `object` types as mentioned in you initial comment.

For now could potentially be a block like
```python
try:
    # pandas >= 1.0 functionality with convert_dtypes
except:
    # fix for pandas < 1.0
```

[Small segment here](https://pandas.pydata.org/pandas-docs/version/1.0.0/user_guide/missing_data.html#missing-data-na-conversion) also makes me think this could be good.

**Comment by fsoubelet** [comment]
With #48 on the way and the shift to pandas `1.0` that comes with it, I think looking into [`convert_dtypes`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.convert_dtypes.html) as i mentioned previously is quite relevant. We could use the method for dtype sanitation when needed (before writing to file?).

I think it's reasonable to adopt the philosophy that if `pandas` itself isn't able to convert to the most appropriate dtype, then there's little hope we can do better.

If we go this route, then after using `.convert_dtypes()` we will inherit `pandas 1.x`-specific types meant to be compatible with `pd.NA`  (as specified in their documentation). It then becomes a matter of properly handling those new dtypes when creating strings to write to file, which I see related to #49 since this check is currently performed with `ID_TO_POSSIBLE_TYPES` and `np.issubdtype` which will play tricks on us for python's default types (`float`, `int` etc) after updating the `numpy` dependency to `1.19`.

Could this work adding a new dictionary similar to `ID_TO_POSSIBLE_TYPES` for the `pandas` types and an additional check against those? Would like your opinion @JoschD @lmalina @mihofer

**Comment by fsoubelet** [comment]
Solved as of commit 929642f (the first one to satisfy codeclimate in #48).

The chosen implementation is to force the use of `pandas` types with `.convert_dtypes(integer=False)` in `write_tfs`.
The types are later on checked by `pandas.api.types`'s own checks instead of `numpy.issubdtype` (which we don't use anymore).

This has minimal impact as it never affects a user's dataframes, since the conversion is done on a copy and only at the time of writing.
